src/main/java/net/minestom/server/instance/Chunk.java

```diff
@@ -40,74 +32,60 @@
-        this.minSection = instanceDim.minY() / CHUNK_SECTION_SIZE;
-        this.maxSection = (instanceDim.minY() + instanceDim.height()) / CHUNK_SECTION_SIZE;
-        final List<SharedInstance> shared = instance instanceof InstanceContainer instanceContainer ?
-                instanceContainer.getSharedInstances() : List.of();
-        this.viewable = instance.getEntityTracker().viewable(shared, chunkX, chunkZ);
+@NotNullByDefault
+public sealed interface Chunk extends Viewable, Tickable, Taggable, Snapshotable permits ChunkImpl {
+    int CHUNK_SIZE_X = SECTION_SIZE;
+    int CHUNK_SIZE_Z = SECTION_SIZE;
+    int CHUNK_SECTION_SIZE = SECTION_SIZE;
+
+    static Chunk chunk(Instance instance, int chunkX, int chunkZ, Consumer<Builder> consumer) {
+        var builder = new ChunkImpl.Builder();
+        consumer.accept(builder);
+        return builder.build(instance, chunkX, chunkZ);
     }
 
-    /**
-     * Sets a block at a position.
-     * <p>
-     * This is used when the previous block has to be destroyed/replaced, meaning that it clears the previous data and update method.
-     * <p>
-     * WARNING: this method is not thread-safe (in order to bring performance improvement with {@link net.minestom.server.instance.batch.Batch batches})
-     * The thread-safe version is {@link Instance#setBlock(int, int, int, Block)} (or any similar instance methods)
-     * Otherwise, you can simply do not forget to have this chunk synchronized when this is called.
-     *
-     * @param x     the block X
-     * @param y     the block Y
-     * @param z     the block Z
-     * @param block the block to place
-     */
-    @Override
-    public void setBlock(int x, int y, int z, @NotNull Block block) {
-        setBlock(x, y, z, block, null, null);
+    static Chunk chunk(Instance instance, int chunkX, int chunkZ) {
+        return new ChunkImpl.Builder().build(instance, chunkX, chunkZ);
+    }
+
+    static Chunk chunkLight(Instance instance, int chunkX, int chunkZ) {
+        return chunk(instance, chunkX, chunkZ, Builder::lightEngine);
+    }
+
+    sealed interface Builder permits ChunkImpl.Builder {
+        void lightEngine(boolean lightEngine);
+
+        default void lightEngine() {
+            lightEngine(true);
+        }
+
+        void readOnly(boolean readOnly);
+
+        default void readOnly() {
+            readOnly(true);
+        }
+
+        void generate(boolean generate);
+
+        default void generate() {
+            generate(true);
+        }
+
+        void sections(List<Section> sections);
     }
 
-    protected abstract void setBlock(int x, int y, int z, @NotNull Block block,
-                                     @Nullable BlockHandler.Placement placement,
-                                     @Nullable BlockHandler.Destroy destroy);
+    List<Section> getSections();
+
+    Section getSection(int section);
 
-    public abstract @NotNull List<Section> getSections();
+    Heightmap motionBlockingHeightmap();
```

Comment:
would be nice to support extra user defined heightmaps - this becomes a lot more annoying to do externally after this change